README
				
				CDPSnarf v0.1.6 Cisco Discovery Protocol Sniffer
				    by Anastasios "Zapotek" Laskos
				        <tasos.laskos@gmail.com>
				          <zapotek@segfault.gr>

	           Licensed under the GNU General Public License v2



General
----------
	CDPSnarf is a network sniffer exclusively written to extract
information from CDP packets.
It provides all the information a "show cdp neighbors detail"
command would return on a Cisco router and even more.

Features
-----------
A feature list follows:
	o Time intervals between CDP advertisements
	o Source MAC address
	o CDP Version
	o TTL
	o Checksum
	o Device ID
	o Software version
	o Platform
	o Addresses
	o Port ID
	o Capabilities
	o Duplex
	o Save packets in PCAP dump file format
	o Read packets from PCAP dump files
	o Debugging information (using the "-d" flag)
	o Tested with IPv4 and IPv6 [1]

[1] Other protocols will be detected and addresses will be printed in an
IPv6 hexadecimal format

Usage
------
Issue the "make" command to built CDPsnarf from source.
Then issue:
	sudo bin/cdpsnarf -i <device>
	
More options with:
	sudo bin/cdpsnarf -h

Requirements
-------------
* PCAP library [http://www.tcpdump.org]
* GCC

Precompiled binaries found under "bin/" are dynamically linked against
libpcap.so.0.8.

Bug reports/Feature requests
-------------------------------
Please send your feedback using GitHub's issue system (http://github.com/Zapotek/cdpsnarf/issues).
I will do my best to apply bugfixes and add new features as fast as possible.

License
---------
CDPSnarf is licensed under the GNU General Public License v2.
See the file "LICENSE" in the same folder for more informaton.

Disclaimer
-----------
CDPSnarf is free software and you are allowed to use it as you see fit.
However, I can't be held responsible for your actions or for any damage
caused by the use of this software.


CHANGELOG

Version 0.1.6
	* Added support for the "IP Prefixes/Gateway" CDP field

Version 0.1.5
	* Added support for dump files (using the "-w <savefile>" argument)
	* Added help output
	* Added IPv6 support
	* Added debugging support (using the "-d" flag)
	* Added source MAC address to the output
	* Added support for extra TLV types
	  (The majority of TLV types is now supported)
	* Added support for reading PCAP dumps instead of listening on an interface
	  (using the "-r <dumpfile>" argument)
	  
Version 0.1.4
	* Improved PCAP filter expression as there were reports of false positives
	  (Props go to the usual suspect...)
	  
Version 0.1.3
	* free'd memory allocations and fixed memory leaks
	* fixed segfault when printing/parsing IP addresses
	(Thanks to paul@expsys.com for pointing that out.)
	
Version 0.1.1
	* invalid IP address returned when CDP packet contained IP address of 0.0.0.0,
	  accompanied by possible segfault.
	  (Thanks to paul@expsys.com for pointing that out.)

includes/oui.h
/*
 *     CDPSnarf CDP packet sniffer
 *   Copyright (C) 2006-2010   Anastasios "Zapotek" Laskos
 *                                  <tasos.laskos@gmail.com>
 *                                  <zapotek@segfault.gr>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License along
 *   with this program; if not, write to the Free Software Foundation, Inc.,
 *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef OUI_H_
#define OUI_H_

/*
 * Registered OUIs: http://standards.ieee.org/regauth/oui/oui.txt
 *
 */

#define OUI_ENCAP_ETHER     0x000000    /* encapsulated Ethernet */
#define OUI_XEROX           0x000006    /* Xerox */
#define OUI_CISCO           0x00000C    /* Cisco (future use) */
#define OUI_NORTEL          0x000081    /* Nortel SONMP */
#define OUI_CISCO_90        0x0000F8    /* Cisco (IOS 9.0 and above?) */
#define OUI_ERICSSON        0x0001EC    /* Ericsson Group */    
#define OUI_CATENA          0x00025A    /* Catena Networks */
#define OUI_SONY_ERICSSON   0x000AD9    /* Sony Ericsson Mobile Communications AB */
#define OUI_SONY_ERICSSON_2 0x000E07    /* Sony Ericsson Mobile Communications AB */
#define OUI_PROFINET        0x000ECF    /* PROFIBUS Nutzerorganisation e.V. */
#define OUI_SONY_ERICSSON_3 0x000FDE    /* Sony Ericsson Mobile Communications AB */
#define OUI_IEEE_802_3      0x00120F    /* IEEE 802.3 */
#define OUI_MEDIA_ENDPOINT  0x0012BB    /* Media (TIA TR-41 Committee) */
#define OUI_SONY_ERICSSON_4 0x0012EE    /* Sony Ericsson Mobile Communications AB */
#define OUI_ERICSSON_MOBILE 0x0015E0    /* Ericsson Mobile Platforms */ 
#define OUI_SONY_ERICSSON_5 0x001620    /* Sony Ericsson Mobile Communications AB */
#define OUI_SONY_ERICSSON_6 0x0016B8    /* Sony Ericsson Mobile Communications AB */
#define OUI_SONY_ERICSSON_7 0x001813    /* Sony Ericsson Mobile Communications AB */
#define OUI_SONY_ERICSSON_8 0x001963    /* Sony Ericsson Mobile Communications AB */
#define OUI_CISCOWL         0x004096    /* Cisco Wireless (Aironet) */
#define OUI_ERICSSON_2      0x008037    /* Ericsson Group */    
#define OUI_BRIDGED         0x0080C2    /* Bridged Frame-Relay, RFC 2427 */
                                        /* and Bridged ATM, RFC 2684 */
#define OUI_IEEE_802_1      0x0080C2    /* IEEE 802.1 Committee */
#define OUI_ATM_FORUM       0x00A03E    /* ATM Forum */
#define OUI_EXTREME         0x00E02B    /* Extreme EDP/ESRP */
#define OUI_CABLE_BPDU      0x00E02F    /* DOCSIS spanning tree BPDU */
#define OUI_SIEMENS         0x080006    /* Siemens AG */
#define OUI_APPLE_ATALK     0x080007    /* Appletalk */
#define OUI_HP              0x080009    /* Hewlett-Packard */

// the names of the above OUI codes
static assoc_array OUI_NAMES[] = {
    { OUI_ENCAP_ETHER,   "Encapsulated Ethernet" },
    { OUI_XEROX,   "Xerox" },
    { OUI_CISCO,   "Cisco" },
    { OUI_NORTEL,   "Nortel SONMP" },
    { OUI_CISCO_90,   "Cisco" },
    { OUI_ERICSSON,   "Ericsson Group" },
    { OUI_CATENA,   "Catena Networks" },
    { OUI_SONY_ERICSSON,   "Sony Ericsson Mobile Communications AB" },
    { OUI_SONY_ERICSSON_2,   "Sony Ericsson Mobile Communications AB" },
    { OUI_PROFINET,   "PROFIBUS Nutzerorganisation e.V." },
    { OUI_SONY_ERICSSON_3,   "Sony Ericsson Mobile Communications AB" },
    { OUI_IEEE_802_3,   "IEEE 802.3" },
    { OUI_MEDIA_ENDPOINT,   "Media (TIA TR-41 Committee)" },
    { OUI_SONY_ERICSSON_4,   "Sony Ericsson Mobile Communications AB" },
    { OUI_ERICSSON_MOBILE,   "Sony Ericsson Mobile Communications AB" },
    { OUI_SONY_ERICSSON_5,   "Sony Ericsson Mobile Communications AB" },
    { OUI_SONY_ERICSSON_6,   "Sony Ericsson Mobile Communications AB" },
    { OUI_SONY_ERICSSON_7,   "Sony Ericsson Mobile Communications AB" },
    { OUI_SONY_ERICSSON_8,   "Sony Ericsson Mobile Communications AB" },
    { OUI_CISCOWL,   "Cisco Wireless (Aironet)" },
    { OUI_ERICSSON_2,   "Ericsson Group" },
    { OUI_BRIDGED,   " Bridged Frame-Relay/ATM" },
    { OUI_IEEE_802_1,   "IEEE 802.1 Committee" },
    { OUI_ATM_FORUM,   "ATM Forum" },
    { OUI_EXTREME,   "Extreme EDP/ESRP" },
    { OUI_CABLE_BPDU,   "DOCSIS spanning tree BPDU" },
    { OUI_SIEMENS,   "Siemens AG" },
    { OUI_APPLE_ATALK,   "Appletalk" },
    { OUI_HP,   "Hewlett-Packard" }
};

#endif /*OUI_H_*/


includes/crc16/crc.c
//#include "crc.h"
//
//const unsigned short crc_table[256] =
//{
//    0x0000, 0x1189, 0x2312, 0x329B, 0x4624, 0x57AD, 0x6536, 0x74BF,
//    0x8C48, 0x9DC1, 0xAF5A, 0xBED3, 0xCA6C, 0xDBE5, 0xE97E, 0xF8F7,
//    0x1081, 0x0108, 0x3393, 0x221A, 0x56A5, 0x472C, 0x75B7, 0x643E,
//    0x9CC9, 0x8D40, 0xBFDB, 0xAE52, 0xDAED, 0xCB64, 0xF9FF, 0xE876,
//    0x2102, 0x308B, 0x0210, 0x1399, 0x6726, 0x76AF, 0x4434, 0x55BD,
//    0xAD4A, 0xBCC3, 0x8E58, 0x9FD1, 0xEB6E, 0xFAE7, 0xC87C, 0xD9F5,
//    0x3183, 0x200A, 0x1291, 0x0318, 0x77A7, 0x662E, 0x54B5, 0x453C,
//    0xBDCB, 0xAC42, 0x9ED9, 0x8F50, 0xFBEF, 0xEA66, 0xD8FD, 0xC974,
//    0x4204, 0x538D, 0x6116, 0x709F, 0x0420, 0x15A9, 0x2732, 0x36BB,
//    0xCE4C, 0xDFC5, 0xED5E, 0xFCD7, 0x8868, 0x99E1, 0xAB7A, 0xBAF3,
//    0x5285, 0x430C, 0x7197, 0x601E, 0x14A1, 0x0528, 0x37B3, 0x263A,
//    0xDECD, 0xCF44, 0xFDDF, 0xEC56, 0x98E9, 0x8960, 0xBBFB, 0xAA72,
//    0x6306, 0x728F, 0x4014, 0x519D, 0x2522, 0x34AB, 0x0630, 0x17B9,
//    0xEF4E, 0xFEC7, 0xCC5C, 0xDDD5, 0xA96A, 0xB8E3, 0x8A78, 0x9BF1,
//    0x7387, 0x620E, 0x5095, 0x411C, 0x35A3, 0x242A, 0x16B1, 0x0738,
//    0xFFCF, 0xEE46, 0xDCDD, 0xCD54, 0xB9EB, 0xA862, 0x9AF9, 0x8B70,
//    0x8408, 0x9581, 0xA71A, 0xB693, 0xC22C, 0xD3A5, 0xE13E, 0xF0B7,
//    0x0840, 0x19C9, 0x2B52, 0x3ADB, 0x4E64, 0x5FED, 0x6D76, 0x7CFF,
//    0x9489, 0x8500, 0xB79B, 0xA612, 0xD2AD, 0xC324, 0xF1BF, 0xE036,
//    0x18C1, 0x0948, 0x3BD3, 0x2A5A, 0x5EE5, 0x4F6C, 0x7DF7, 0x6C7E,
//    0xA50A, 0xB483, 0x8618, 0x9791, 0xE32E, 0xF2A7, 0xC03C, 0xD1B5,
//    0x2942, 0x38CB, 0x0A50, 0x1BD9, 0x6F66, 0x7EEF, 0x4C74, 0x5DFD,
//    0xB58B, 0xA402, 0x9699, 0x8710, 0xF3AF, 0xE226, 0xD0BD, 0xC134,
//    0x39C3, 0x284A, 0x1AD1, 0x0B58, 0x7FE7, 0x6E6E, 0x5CF5, 0x4D7C,
//    0xC60C, 0xD785, 0xE51E, 0xF497, 0x8028, 0x91A1, 0xA33A, 0xB2B3,
//    0x4A44, 0x5BCD, 0x6956, 0x78DF, 0x0C60, 0x1DE9, 0x2F72, 0x3EFB,
//    0xD68D, 0xC704, 0xF59F, 0xE416, 0x90A9, 0x8120, 0xB3BB, 0xA232,
//    0x5AC5, 0x4B4C, 0x79D7, 0x685E, 0x1CE1, 0x0D68, 0x3FF3, 0x2E7A,
//    0xE70E, 0xF687, 0xC41C, 0xD595, 0xA12A, 0xB0A3, 0x8238, 0x93B1,
//    0x6B46, 0x7ACF, 0x4854, 0x59DD, 0x2D62, 0x3CEB, 0x0E70, 0x1FF9,
//    0xF78F, 0xE606, 0xD49D, 0xC514, 0xB1AB, 0xA022, 0x92B9, 0x8330,
//    0x7BC7, 0x6A4E, 0x58D5, 0x495C, 0x3DE3, 0x2C6A, 0x1EF1, 0x0F78
//};
//
//
//unsigned short crc16_checksum(char * buf,unsigned long length)
//{
//	unsigned long i;
//	unsigned short crc=0xffff;
//
//	for(i=0;i<length;i++) {
//		CRC(crc,buf[i]);
//	}
//	
//	return crc;
//}
//
//
//void crc16_append(char * buf, unsigned long length)
//{
//	unsigned short crc = crc16_checksum(buf,length);
//	buf[length] = (unsigned char)(crc&0xff);
//	buf[length+1] = (unsigned char)((crc>>8) & 0xff);
//}

/*
**************************************************************************
Function: ip_sum_calc
Description: Calculate the 16 bit IP sum.
***************************************************************************
*/
typedef unsigned short u16;
typedef unsigned long u32;

u16 crc16_checksum(u_char buff[], u16 len_ip_header)
{
u16 word16;
u32 sum=0;
u16 i;
    
    // make 16 bit words out of every two adjacent 8 bit words in the packet
    // and add them up
    for (i=0;i<len_ip_header;i=i+2){
        word16 =((buff[i]<<8)&0xFF00)+(buff[i+1]&0xFF);
        sum = sum + (u32) word16;   
    }
    
    // take only 16 bits out of the 32 bit sum and add up the carries
    while (sum>>16)
      sum = (sum & 0xFFFF)+(sum >> 16);

    // one's complement the result
    sum = ~sum;
    
return ((u16) sum);
}


cdpsnarf.h
/*
 *     CDPSnarf CDP packet sniffer
 *   Copyright (C) 2006-2010   Anastasios "Zapotek" Laskos
 *                                  <tasos.laskos@gmail.com>
 *                                  <zapotek@segfault.gr>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License along
 *   with this program; if not, write to the Free Software Foundation, Inc.,
 *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef SNARF_H_
#define SNARF_H_

typedef struct _assoc_array {
  int   value;
  const char   *string;
} assoc_array;

typedef unsigned short uint16_t;
typedef unsigned int uint32_t;

#include "includes/oui.h"
//#include "includes/crc16/crc.c"

struct address {
    int proto_type;
    int proto_len;
    int proto;
    int address_len;
    int address;
} *addresses;

// Type codes of the CDP TLV
#define TYPE_DEVICE_ID      0x0001 // supported
#define TYPE_ADDRESS        0x0002 // supported
#define TYPE_PORT_ID        0x0003 // supported
#define TYPE_CAPABILITIES   0x0004 // supported
#define TYPE_IOS_VERSION    0x0005 // supported
#define TYPE_PLATFORM       0x0006 // supported
#define TYPE_IP_PREFIX      0x0007 // supported but needs further testing
#define TYPE_PROTOCOL_HELLO     0x0008 // supported
#define TYPE_VTP_MGMT_DOMAIN    0x0009 // supported
#define TYPE_NATIVE_VLAN        0x000a // supported
#define TYPE_DUPLEX             0x000b // supported
/*                              0x000c */
/*                              0x000d */
#define TYPE_VOIP_VLAN_REPLY    0x000e
#define TYPE_VOIP_VLAN_QUERY    0x000f
#define TYPE_POWER              0x0010
#define TYPE_MTU                0x0011 // supported
#define TYPE_TRUST_BITMAP       0x0012 // supported
#define TYPE_UNTRUSTED_COS      0x0013 // supported
#define TYPE_SYSTEM_NAME        0x0014 // supported
#define TYPE_SYSTEM_OID         0x0015 // supported
#define TYPE_MANAGEMENT_ADDR    0x0016 // supported
#define TYPE_LOCATION           0x0017
#define TYPE_EXT_PORT_ID        0x0018
#define TYPE_POWER_REQUESTED    0x0019
#define TYPE_POWER_AVAILABLE    0x001a // not fully supported
#define TYPE_PORT_UNIDIR        0x001b

// the names of the above type codes in the same order
static const char* TYPE_NAMES[] = {
    NULL,
    "Device ID",
    "Addresses" ,
    "Port ID" ,
    "Capabilities" ,
    "Software version" ,
    "Platform" ,
    "IP Prefix/Gateway (used for ODR)" ,
    "Protocol Hello" ,
    "VTP Management Domain" ,
    "Native VLAN" ,
    "Duplex" ,
    NULL,
    NULL,
    "VoIP VLAN Reply" ,
    "VoIP VLAN Query" ,
    "Power consumption" ,
    "MTU",
    "Trust Bitmap" ,
    "Untrusted Port CoS" ,
    "System Name" ,
    "System Object ID" ,
    "Management Address" ,
    "Location" ,
    "External Port-ID" ,
    "Power Requested" ,
    "Power Available" ,
    "Port Unidirectional" ,
};

#define TYPE_HELLO_CLUSTER_MGMT    0x0112

static assoc_array type_hello_vals[] = {
        { TYPE_HELLO_CLUSTER_MGMT,   "Cluster Management" },
        { 0,                    NULL }
};

#define MAC_OFFSET 6

// bytes the from the beggining of the packet used for Ethernet and LLC
#define ENCAP_OFFSET  22

// sizes of the type and length fields in the TLV structure
#define TLV_TYPE_SIZE     2
#define TLV_LENGTH_SIZE   2


// layer 2 protocol type available
char *PROTO_TYPES[] = {
    "Uknown",
    "NLPID",
    "802.2"
};

// IDs of Layer 3 protocols
// NOTE: Keep these in order
double long PROTO[] = {
    0x81,
    0xCC,
    0x86DD, // Cisco says 0x0800 for IPv6 alhtough it is 86DD
    0x6003,
    0x809B,
    0x8137,
    0x80c4,
    0x0600,
    0x8019
};

// Layer 3 protocols used
// NOTE: Keep these in order
char PROTO_NAMES[][15] = {
    "ISO CLNS",
    "IP",
    "IPv6",
    "DECNET Phase IV",
    "AppleTalk",
    "Novell IPX",
    "Banyan VINES",
    "XNS",
    "Apollo Domain"
};

// bit masks for each available capability
int CAPABILITIES[] = {
    0x01,
    0x02,
    0x04,
    0x08,
    0x10,
    0x20,
    0x40
};

// capability names
char CAPABILITIES_NAMES[7][19] = {
    "Router",
    "Transparent bridge",
    "Source Route Bridge",
    "Switch",
    "Host",
    "IGMP",
    "Repeater"
};

#endif /*SNARF_H_*/
